sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -151,6 +164,30 @@ abstract class Expression extends TreeNode[Expression] {
   }
 
   /**
+   * Returns the hash for this expression. Expressions that compute the same result, even if
+   * they differ cosmetically should return the same hash.
+   */
+  def semanticHash() : Int = {
```

Comment:
Since all of the expressions are the case class, probably we don't need to our own way to `computeHash`.